nvsetvol.8
.TH NVSETVOL 8 "16th February 2003"

.SH NAME
nvsetvol \- tool for changing startup volume of PowerMac machines

.SH SYNOPSIS
.B nvsetvol
[
.B volume
]

.SH DESCRIPTION
.B nvsetvol
queries and sets the base volume of a PowerMac machine.  With no options,
.B nvsetvol
displays the current volume; with a numerical argument, it sets the volume.
The
.B volume
argument should be a number from 0 to 255.  0 will turn off the annoying
startup chime.

.SH AUTHOR
.B nvsetvol
was written by Matteo Frigo <athena@fftw.org>.

This manual page was written by Daniel Jacobowitz <dan@debian.org> for the
Debian GNU/Linux distribution.


debian/changelog
pmac-utils (1.1.3-27+ports) unreleased; urgency=medium

  * Upload to unreleased for powerpc and ppc64.
  * Bump compat level to 9 in debian/compat.
  * Drop #define __USE_GNU from nwnvsetenv.c.
  * Fix syntax error in autoboot.sgml.

 -- John Paul Adrian Glaubitz <glaubitz@physik.fu-berlin.de>  Tue, 08 Jan 2019 17:48:12 +0300

pmac-utils (1.1.3-27) unstable; urgency=medium

  * Incremental packaging updates for Standards-Version: 3.9.7:
    - Record our source format (1.0) in debian/source/format.
    - Stop redundantly setting DH_COMPAT in debian/rules.
    - Add build-indep and build-arch targets to debian/rules.
    - Use dh_prep instead of dh_clean -k in debian/rules.
    - Add comment-only watch file to document upstream state.
    - Use ${misc:Depends} in debian/control for binaries.
    - Remake debian/copyright with a DEP-5 compatible format.

 -- Adam Conrad <adconrad@0c3.net>  Tue, 05 Apr 2016 12:48:28 -0600

pmac-utils (1.1.3-26) unstable; urgency=medium

  * Add myselt to Uploaders with Michael's consent.
  * Rename to pmac-utils, powerpc-utils is another project.
  * Stop building on little-endian arches (LP: #1398163)
  * Drop ancient and obsolete Conflicts/Replaces.

 -- Adam Conrad <adconrad@0c3.net>  Sun, 27 Mar 2016 13:42:07 -0600

powerpc-utils (1.1.3-25) unstable; urgency=medium

  * Update maintainer address. Closes: #762790.
  * Fix compiler warnings.

 -- Michael Schmitz <schmitzmic@gmail.com>  Wed, 29 Oct 2014 20:28:32 +1100

powerpc-utils (1.1.3-24.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Add ppc64el to the Architecture: field. Closes: #738140.

 -- Aurelien Jarno <aurel32@debian.org>  Sat, 20 Sep 2014 22:09:45 +0200

powerpc-utils (1.1.3-24) unstable; urgency=low

  * Fix fixed postinstall handling of adjtime setup (copy /etc/adjtime if it
    exists, create new file with the correct data otherwise). 
    Closes: #528443.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 13 May 2009 20:04:24 +1100

powerpc-utils (1.1.3-23) unstable; urgency=low

  * Fix broken -f option for fnset. Check for presence of ADB devices and bail 
    out on USB based machines.
    Closes: #474814.

  * Fix postinstall handling of adjtime setup (copy /etc/adjtime if it
    exists, create new file otherwise). 
    Closes: #528443.

  * Update bootsched manpage to clarify that it should only be used on 
    Oldworld PowerMacs.
    Closes: #407295.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 13 May 2009 19:34:24 +1100

powerpc-utils (1.1.3-22) unstable; urgency=low

  * Fix problem with nvram reads segfaulting on efika, reported by Robert
    Millan <rm@aybabtu.com> and patched by same.
    Closes: #431116.

  * Honor DEB_BUILD_OPTS=noopt, patch by Robert Millan <rm@aybabtu.com>.
    Closes: #431117.

  * Add Aurelien Gerome <ag@roxor.cx> as co-maintainer.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed,  4 Jul 2007 17:21:40 +0100

powerpc-utils (1.1.3-21) unstable; urgency=low

  * Fix problem with truncated nvram reads/writes, reported by Adam
    Lackorzynski <adam@os.inf.tu-dresden.de>.
    Closes: #413131.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed,  7 Mar 2007 15:57:34 +0100

powerpc-utils (1.1.3-20) unstable; urgency=low

  * Change section to 'utils' to match override file.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Tue, 30 Jun 2006 15:39:34 +0100

powerpc-utils (1.1.3-19) unstable; urgency=low

  * The 'upload lost' release :-(

  * Incorporate newworld autoboot patch by Rich Johnson.

  * Fix description formatting (thanks to Nick Boullis <nboullis@debian.org>
    for the alert).
    Closes: #336972.

  * Add ppc64 arch (just in case biarch powerpc isn't enough).
    Closes: #360797.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Tue, 30 Jun 2006 15:39:34 +0100

powerpc-utils (1.1.3-18) unstable; urgency=low

  * add -D_GNU_SOURCE to Makefile in order to help with 64 bit pointer
    size on ppc64.
    Closes: #326394.

  * Fix problem introduced in -12 (/etc/adjtime converted to symlink 
    into /var fs): Convert symlink to plain copy now.
    Closes: #332926.

  * Fix char signedness bug in nvsetvol (patch by Eugen Dedu
    <Eugen.Dedu@pu-pm.univ-fcomte.fr>.
    Closes: #334547

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 19 Oct 2005 18:07:34 +0200

powerpc-utils (1.1.3-17) unstable; urgency=low

  * Fix for iBook G4 failure to save nvram data (as manifest by failure to
    properly store the boot chime volume). Patch by Joerg Dorchain
    <joerg@dorchain.net>.
    Closes: #321655, #320566.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Mon, 08 Aug 2005 14:40:34 +0200

powerpc-utils (1.1.3-16) unstable; urgency=low

  * Brown Paper Bag release: Added bootsched util (for CUDA based Powermacs)
    the source for which I had buried for over three years now.

  * Finally, actually install the autoboot binary (reported missing by
    Herve Eychenne <rv@eychenne.org>).
    Closes: #317021.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 06 Jul 2005 09:42:44 +0200

powerpc-utils (1.1.3-15) unstable; urgency=low

  * Merge back patches from ubuntu (thanks to Adam Conrad <adconrad@0c3.net>
    for the hint): fix linuxdoc-tools b-d, drop adb_mouse.h header. Update 
    description to include autoboot and lsprop tools.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Mon, 09 May 2005 12:31:24 +0100

powerpc-utils (1.1.3-14) unstable; urgency=low

  * Fixed nvsetvol failure (infinite loop reading NVRAM) on OldWorld
    machines (reported by Christophe Donatsch <cdonatsch@tele2.ch>).
    Closes: #304493.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Sun, 01 May 2005 11:58:37 +0100

powerpc-utils (1.1.3-13) unstable; urgency=low

  * Fixed nvsetenv's devfsd bug (reported by Roger Leigh <rleigh@debian.org>).
    Closes: #300946.

  * Added autoboot util (for CUDA based Powermacs)


nvsetvol.c
/* set volume of the startup chime in the PowerMac nvram. 

   Written by Matteo Frigo <athena@fftw.org>.
   $Id: nvsetvol.c,v 1.1 2003/01/11 11:26:30 athena Exp $

   This program is in the public domain.
*/

/* The volume is stored as a byte at address 8 in the parameter ram. */

#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <asm/nvram.h>
#ifndef IOC_NVRAM_SYNC
#warning IOC_NVRAM_SYNC undefined -- update your headers
#define IOC_NVRAM_SYNC _IO('p', 0x43)
#endif

typedef struct {
     unsigned char sig;
     unsigned char cksum;
     unsigned short len;
     char name[12];
} header;

void die(const char *s) __attribute__((noreturn));
void die(const char *s)
{
     perror(s);
     exit(1);
}

void seek_pram(int fd, int offset)
{
     if (lseek(fd, offset, SEEK_SET) < 0)
	  die("lseek");
}

// MSch: only works for newworld - oldworld have XPRAM at offset 0x1300 fixed. 

int find_pram(int fd, int *size)
{
     header buf;
     int offset = 0;
     int rc = 0;
     while((rc = read(fd, &buf, sizeof(header))) == sizeof(header)) {
	  int sz = buf.len * 16;
	  // what's the sig in char?
	  // if (buf.sig == 0x1275) fprintf(stderr, "sig at offset %d rc %d buf.len %d buf.name %s \n", offset, rc, buf.len, buf.name);
	  // if (buf.sig == 0x5a) fprintf(stderr, "sig at offset %x rc %d buf.len %x buf.name %s \n", offset, rc, buf.len, buf.name);
	  // if (sz) fprintf(stderr, "offset %x rc %d sig %x buf.len %x buf.name %s \n", offset, rc, buf.sig, buf.len, buf.name);
	  if (!strncmp(buf.name, "APL,MacOS75", 11)) {
	       *size = sz - sizeof(header);
	       // fprintf(stderr, "XPRAM offset %x size %x\n", offset + sizeof(header), *size); 
	       return offset + sizeof(header);
	  }
	  if (sz)
	    offset += sz;
          else
            offset += sizeof(header);
	  seek_pram(fd, offset);
     }
     // no Core99 style PRAM found - just assume hard coded values as set in kernel:
     // XPRAM 0x1300
     // NR    0x1400
     // OF    0x1800
     // TODO: use ioctl to get PRAM offset
     // fprintf(stderr, "no NW PRAM found, assuming OW XPRAM offset 0x1300 size 0x100\n");
     *size = 0x100;
     return (0x1300);
     // NOTREACHED
     die("no PRAM found");
}

#define VOLADDR 8

int main(int argc, char *argv[])
{
     int fd, offset, size;
     char *buf;

     fd = open("/dev/nvram", O_RDWR);
     if (fd < 0) die("can't open /dev/nvram");
     
     offset = find_pram(fd, &size);
     buf = malloc(size);
     if (!buf) die("can't malloc()");

     seek_pram(fd, offset);
     if (read(fd, buf, size) != size)
	  die("error reading /dev/nvram");

     printf("current volume is %d\n", (unsigned char) buf[VOLADDR]);


#ifdef DEBUG_PRAM
     for (i=0; i < size; i++) {
          printf("cell %d : %d\n", i, (unsigned char) buf[i]);
     }
#endif
     
     if (argc > 1) {
	  buf[VOLADDR] = atoi(argv[1]);

	  seek_pram(fd, offset);
	  if (write(fd, buf, size) != size)
	       die("error writing /dev/nvram");
	  printf("new volume is %d\n", (unsigned char) buf[VOLADDR]);
     }
     ioctl(fd, IOC_NVRAM_SYNC);
     close(fd);
     return 0;
}
